File: src/main/kotlin/sunsetsatellite/vm/sunlite/CallFrame.kt
package sunsetsatellite.vm.sunlite

import com.sun.org.apache.xalan.internal.lib.ExsltDynamic.closure
import java.util.*

class CallFrame(val closure: SLClosure, val locals: MutableList<AnySLValue>) {
	var pc: Int = 0
	val stack: Stack<AnySLValue> = Stack()

	fun pop(): AnySLValue {
		return stack.pop()
	}

	fun push(value: AnySLValue) {
		stack.push(value)
	}

	fun peek(): AnySLValue {
		return stack.peek()
	}

	fun peek(i: Int): AnySLValue {
		val len = stack.size

		if (len == 0) throw EmptyStackException()
		return stack.elementAt(len - i - 1)
	}

	override fun toString(): String {
		return "[line ${closure.function.chunk.debugInfo.lines[pc]}] in ${if(closure.function.name == "") "script" else "${closure.function.name}()"}"
	}
}

File: src/main/csharp/SunliteSharp/Runtime/CallFrame.cs
namespace SunliteSharp.Runtime;

public record CallFrame(SLClosure Closure, List<AnySLValue> Locals)
{
    public int Pc = 0;
    public List<AnySLValue> Stack = [];

    public AnySLValue Pop()
    {
        var value = Stack[^1];
        Stack.RemoveAt(Stack.Count - 1);
        return value;
    }
    
    public void Push(AnySLValue value)
    {
        Stack.Add(value);
    }

    public AnySLValue Peek()
    {
        return Stack[^1];
    }

    public AnySLValue Peek(int offset)
    {
        return Stack[Stack.Count - offset - 1];
    }

    public override string ToString()
    {
        return $"[line {Closure.Function.Chunk.Header.Lines[Pc]}] in {(Closure.Function.Name == "" ? $"file {Closure.Function.Chunk.Header.File}" : $"function {Closure.Function.Name}")}";
    }
}